src/emmy/complex.cljc

```diff
@@ -7,167 +7,82 @@
 
-#?(:clj
-   (def complex-format (ComplexFormat.)))
-
 (defn complex
   "Returns a [[Complex]] number with the supplied real part `re` and imaginary
   part `im`. `im` defaults to 0."
-  ([re]
-   #?(:clj (if (string? re)
-             (.parse ^ComplexFormat complex-format re)
-             (Complex. (u/double re)))
-      :cljs (Complex.
-             (if (string? re)
-               re
-               (u/double re)))))
+  ([c]
+   (cond (string? c) (c/parse c)
+         (vector? c) (do
+                       (assert (= (count c) 2))
+                       (let [[re im] c] (c/->Complex re im)))
+         :else (c/->Complex c 0)))
   ([re im]
-   (Complex. (u/double re)
-             (u/double im))))
+   (c/->Complex re im)))
+
+(defn parse-complex
+  "Used as a [data reader](https://www.clojurescript.org/guides/reader#_clojurescript_compilation).
+  In Clojure, we could get away with just using the constructor above, but in
+  ClojureScript we need to act as a macro to delay evaluation of the constructor."
+  [c]
+  `(complex ~c))
 
 (defn complex?
   "Returns true if `a` is an instance of [[Complex]], false otherwise."
   [a]
-  (instance? Complex a))
-
-(defn ^:no-doc real [^Complex a]
-  #?(:clj (.getReal a)
-     :cljs (obj/get a "re")))
-
-(defn ^:no-doc imaginary [^Complex a]
-  #?(:clj (.getImaginary a)
-     :cljs (obj/get a "im")))
-
-(defn ^:no-doc parse-complex
-  "Parser that converts a string, vector or numeric representation of a complex
-   number, like
+  (instance? emmy.complex.impl.Complex a))
 
-  - `1 + 3i`
-  - [1 3]
-  - 1
-
-  into a [[Complex]] number object in clj or cljs."
-  [x]
-  (cond (string? x)
-        #?(:clj
-           (let [v (.parse ^ComplexFormat complex-format x)]
-             `(complex ~(real v) ~(imaginary v)))
-           :cljs `(complex ~x))
-
-        (vector? x)
-        (let [[re im] x]
-          (if (nil? im)
-            `(complex ~re)
-            `(complex ~re ~im)))
-
-        (number? x) `(complex ~x)
+(defn ^:no-doc real
+  [z]
+  (c/real z))
 
-        :else (u/illegal
-               (str
-                "#emmy/complex takes a string, 2-vector or a number. Received: "
-                x))))
+(defn ^:no-doc imaginary
+  [z]
+  (c/imaginary z))
 
 ;; ## Type Extensions
 
-#?(:cljs
-   (extend-type Complex
-     IEquiv
-     (-equiv [this other]
-       (cond (complex? other)
-             (.equals ^js this other)
-
-             (v/real? other)
-             (and (zero? (imaginary this))
-                  (v/= (real this) other))
-
-             ;; Defer to `v/=` to support quaternion, octonion equality etc.
-             :else (v/= this other)))
-
-     IPrintWithWriter
-     (-pr-writer [x writer _]
-       (write-all
-        writer
-        "#emmy/complex "
-        (str [(obj/get x "re")
-              (obj/get x "im")])))))
-
 #?(:clj
-   (defmethod print-method Complex [^Complex v ^java.io.Writer w]
+   (defmethod print-method emmy.complex.impl.Complex [v ^java.io.Writer w]
```

Comment:
ah, I see you have this here already. I might recommend moving it next to the type, so if someone only requires `emmy.complex.impl` they get this behavior too.

Reply:
I agree/fixing